### PR TITLE
Teach tasks what machine they are running on

### PR DESCRIPTION
**Pull request by jihoonson**
Part of https://github.com/apache/incubator-druid/issues/8061.

### Description

In parallel native indexing, the middleManager is responsible for serving intermediary data for shuffle. As a result, the phase 2 tasks need to know where phase 1 tasks ran.

This PR lets tasks know where they are running by adding new system properties. This is supposed to be used for only internal purpose and so doesn't have to be documented.

<hr>

This PR has:
- [x] been self-reviewed.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: core/src/main/java/org/apache/druid/guice/annotations/Parent.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * Indicates that the DruidNode bound with this annotation holds the information of the machine which spawned
+ * this process.
+ *
+ * @see Self
+ */
+@Target({ElementType.FIELD, ElementType.PARAMETER})
+@Retention(RetentionPolicy.RUNTIME)
+@BindingAnnotation
+@PublicApi
+public @interface Parent
```
Comment (jon-wei) [inline core/src/main/java/org/apache/druid/guice/annotations/Parent.java]: suggest `ParentNode` instead to make it more clear what the annotation relates to

Comment (jihoonson) [reply core/src/main/java/org/apache/druid/guice/annotations/Parent.java]: Hmm I want to make it consistent with `Self`. Do you want me to rename it as well? I guess javadoc is pretty clear now that it relates to `DruidNode`.

Comment (jon-wei) [reply core/src/main/java/org/apache/druid/guice/annotations/Parent.java]: Hm, that's fine then, no need to rename in this patch

**Comment by fjy** [comment]
@jihoonson can you fix merge conflicts?

**Comment by jihoonson** [comment]
@fjy fixed.

**Comment by jihoonson** [comment]
@clintropolis thanks, fixed.

**Pull request merged**